aspnetcore/fundamentals/ihostedservice.md

```diff
@@ -0,0 +1,78 @@
+ms.topic: article
+uid: fundamentals/ihostedservice
+---
+# Background tasks with IHostedService in ASP.NET Core
+
+By [Luke Latham](https://github.com/guardrex)
+
+In ASP.NET Core, background tasks are implemented as *hosted services*. A hosted service is a class with background task logic that implements the [IHostedService](/dotnet/api/microsoft.extensions.hosting.ihostedservice) interface. This topic provides three hosted service examples:
+
+* Background task that runs on a timer.
+* Hosted service that activates a scoped service. The scoped service can use dependency injection.
+* Queued background tasks that run sequentially.
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/ihostedservice/samples/) ([how to download](xref:tutorials/index#how-to-download-a-sample))
+
+## IHostedService interface
+
+Hosted services implement the [IHostedService](/dotnet/api/microsoft.extensions.hosting.ihostedservice) interface. The interface defines two methods for objects that are managed by the host:
+
+* [StartAsync(CancellationToken)](/dotnet/api/microsoft.extensions.hosting.ihostedservice.startasync) - Called after the server has started and [IApplicationLifetime.ApplicationStarted](/dotnet/api/microsoft.aspnetcore.hosting.iapplicationlifetime.applicationstarted) is triggered. `StartAsync` contains the logic to start the background task.
+
+* [StopAsync(CancellationToken)](/dotnet/api/microsoft.extensions.hosting.ihostedservice.stopasync) - Triggered when the host is performing a graceful shutdown. `StopAsync` contains the logic to end the background task and dispose of any unmanaged resources.
+
+The hosted service is activated once at app startup and gracefully shutdown at app shutdown.
+
+## Timed background tasks
+
+A timed background task makes use of the [System.Threading.Timer](/dotnet/api/system.threading.timer) class. The timer triggers the task's work and is disposed when the service stops:
+
+[!code-csharp[](ihostedservice/sample/Services/TimedHostedService.cs?name=snippet1&highlight=9,23)]
+
+The service is registered in `Startup.ConfigureServices`:
+
+[!code-csharp[](ihostedservice/sample/Startup.cs?name=snippet1)]
+
+## Consuming a scoped service in a background task
+
+Since an `IHostedService` is started when the app starts, the service is activated only *once*. To use dependency injection with a hosted service, a scope must be created during execution. The scope is used resolve an independently registered service where dependency injection can be used.
+
+The scoped background task service contains the background task's logic. In the following example, [IHostingEnvironment](/dotnet/api/microsoft.aspnetcore.hosting.ihostingenvironment) is injected into the service:
+
+[!code-csharp[](ihostedservice/sample/Services/ScopedProcessingService.cs?name=snippet1&highlight=8,12,18)]
+
+The hosted service creates a scope to resolve the scoped background task service to call its `DoWork` method:
```

Comment:
The ScopedHostedService name is confusing, the hosted service isn't scoped, it consumes a scoped service. The example is strange too. If this were the timer example the scope would be created and destroyed inside of DoWork, not in StartAsync.

Reply:
I was trying to name them in a way that pertains to what they're demonstrating; and yes, *it went horribly wrong here.* The reason the scope creation is in `StartAsync` is because your sample seemed to want it there: See `StartAsync` at https://github.com/aspnet/Docs/issues/3352#issuecomment-333254142. I'll refactor and work on the naming for the next commit.

Reply:
Fair enough, my sample worked but wasn't sensible.